pubsub/iterator.go

```diff
@@ -598,13 +690,73 @@ func (it *messageIterator) sendAck(m map[string]*AckResult) {
 // percentile in order to capture the highest amount of time necessary without
 // considering 1% outliers. If the ModAck RPC fails and exactly once delivery is
 // enabled, we retry it in a separate goroutine for a short duration.
-func (it *messageIterator) sendModAck(m map[string]*AckResult, deadline time.Duration, logOnInvalid bool) {
+func (it *messageIterator) sendModAck(m map[string]*AckResult, deadline time.Duration, logOnInvalid, isReceipt bool) {
 	deadlineSec := int32(deadline / time.Second)
-	it.sendAckWithFunc(m, func(ctx context.Context, subName string, ackIds []string) error {
+	isNack := deadline == 0
+	var eventStart, eventEnd string
+	if isNack {
+		eventStart = eventNackStart
+		eventEnd = eventNackEnd
+	} else {
+		eventStart = eventModackStart
+		eventEnd = eventModackEnd
+	}
+	it.sendAckWithFunc(m, func(ctx context.Context, subName string, ackIDs []string) error {
+		links := make([]trace.Link, 0, len(ackIDs))
+		subscribeSpans := make([]trace.Span, 0, len(ackIDs))
+		if it.enableTracing {
+			for _, ackID := range ackIDs {
+				// get the parent span context for this ackID for otel tracing.
+				var s any
+				var ok bool
+				if isNack {
+					s, ok = it.activeSpans.LoadAndDelete(ackID)
+				} else {
+					s, ok = it.activeSpans.Load(ackID)
+				}
+				if ok {
+					subscribeSpan := s.(trace.Span)
+					subscribeSpans = append(subscribeSpans, subscribeSpan)
+					if isNack {
+						defer subscribeSpan.End()
+						defer subscribeSpan.SetAttributes(attribute.String(resultAttribute, resultNacked))
+					}
+					subscribeSpan.AddEvent(eventStart, trace.WithAttributes(semconv.MessagingBatchMessageCount(len(ackIDs))))
+					defer subscribeSpan.AddEvent(eventEnd)
+					links = append(links, trace.Link{SpanContext: subscribeSpan.SpanContext()})
+				}
+			}
+		}
+		var mSpan trace.Span
```

Comment:
I wonder if you could just put this `var` inside the `if` to make one long `enableTracing`.